### DISCUSSION TITLE
Is workerd intended to be resilient in the face of untrusted HTTP clients?

### DISCUSSION THREAD

**Discussion by isker** [cloudflare/workerd#1238] [Q&A] [answered]
Hi! I'm mostly just curious (we will use it behind another trusted L7 proxy), but is workerd intended to be able to serve as an "edge" ingress proxy where it is taking traffic from untrusted clients?

For what it's worth, I did a brief assessment on one class of misbehaviors, using [slowhttptest](https://github.com/shekyan/slowhttptest/wiki), which can run "slow" dos attacks, e.g. slowloris (neverending http headers), slow POST (neverending body), etc.  As far as I can tell, slowloris is mitigated (seems there is a fixed 15s timeout on receiving all the headers somewhere, though I wasn't able to find it, looking briefly), and it is possible to implement mitigations against body attacks by timing out stream operations in JavaScript, though there's no direct support from workerd in this regard; there is some warning/error output when you stop reading a Request body stream and return a Response, but it didn't seem to result in any actual server degradation.

Regardless of what the answer is, it may be worth clearly stating it in the README or somewhere else in the documentation.

**Comment by kentonv** [comment, marked as answer]
Yes, workerd is intended to be able to handle malicious direct clients. That said, in Cloudflare's production usage, it does not do so -- workerd sits behind Cloudflare's usual proxy stack which is probably more hardened. workerd's HTTP implementation comes from KJ, a library bundled with Cap'n Proto, which I personally wrote. It is admittedly not widely used outside of Cloudflare Workers.

workerd uses an event-driven architecture where the resources allocated to each connection are fairly minimal, so it should have no problem with "slow loris" attacks. That said, the application must play a part in this too: if the application allocates a large amount of state before fully reading the request body, this could lead to a vulnerability. In production on Cloudflare Workers, our usual proxy stack (outside of workerd) buffers requests and does not deliver them until the entire body is received; workerd itself does not implement any such mitigation and leaves it up to the app or a reverse proxy.

Returning a repsonse without reading a whole request should work fine. There is code in KJ that will attempt to read and discard the rest of the request in order to allow connection reuse, but it applies both a time limit and a byte limit to this.

**Reply by isker** [reply]
Thanks! All very interesting information.

> but it applies both a time limit and a byte limit to this.

Are these limits parameterized in KJ? If so, do you think they could be configurable in workerd?

**Reply by kentonv** [reply]
This is configurable at the KJ level:

https://github.com/capnproto/capnproto/blob/3ca87c135c98539816730d3c5bff479140d5e6e8/c++/src/kj/compat/http.h#L1175-L1181

I suppose we could expose these options via workerd config, but personally I've never remotely felt a desire to tune these options. I'd be curious to understand why you're interested in them.

**Reply by isker** [reply]
Thanks for the link.

> I'd be curious to understand why you're interested in them.

Like I said, I'm just curious about the "programmable proxy" angle of workerd.  I don't actually need to tune them myself as we'll be behind another reverse proxy (where we do in fact tune these kinds of options, though mostly with respect to http body streams).

Thanks for your help!
